packages/block-editor/src/store/actions.js

```diff
@@ -1669,14 +1669,16 @@ export const setNavigationMode =
  */
 export const __unstableSetEditorMode =
 	( mode ) =>
-	( { registry } ) => {
+	( { registry, dispatch } ) => {
 		registry.dispatch( preferencesStore ).set( 'core', 'editorTool', mode );
 
 		if ( mode === 'navigation' ) {
 			speak( __( 'You are currently in Write mode.' ) );
 		} else if ( mode === 'edit' ) {
 			speak( __( 'You are currently in Design mode.' ) );
 		}
+
+		dispatch( { type: 'SET_EDITOR_MODE', mode } );
```

Comment:
This is the one that I'm not sure about because some people could call `registry.dispatch( preferencesStore ).set( 'core', 'editorTool', mode );` directly. Maybe it's fine I don't know

Reply:
Yeah, that's a good point, though I guess the spoken messages also won't work when changing the mode that way. Some options:
- (I think you mentioned this one) Always calculate the block editing modes for write mode even when in design mode, then the selector can choose the right one. This'd be a shame as I expect it'd cause a performance hit.
- Add a listener (probably an effect) that reacts to the preference change and sets the mode in the block editor store.
- Make the preference key private so that it can only be called internally. I think this preference has already shipped in the plugin, so it would need some back compat wrangling. Possibly my preferred solution though.

Reply:
Also, what happens on "init"? which mode is used?

I may be leaning towards the first option personally to decouple the preference setting entirely from the computation of the mode.

Reply:
> Also, what happens on "init"? which mode is used?

It uses whatever value the preferences store has at the time of computation. I guess there could be a concern if the preferences store isn't ready yet for some reason.

Reply:
Apart from resolving this question, the rest of the PR is pretty finalized (I've added lots of tests).

I'll have a go at your suggestion, I think it should be pretty easy to implement, we'll probably just need an extra state property for `navModeBlockEditingModes` or something. One thing on my mind is preventing this from being calculated when write mode isn't an option (I think in the post editor it's not an option), as that would be unnecessary.